From 313b79414b5dd2cb8ee8ff3de3b6a55bcaafc466 Mon Sep 17 00:00:00 2001
From: Vinu Vaghasia <vinu.vaghasia@amd.com>
Date: Thu, 28 Apr 2022 18:23:23 -0500
Subject: [PATCH 1/1] u-boot-aspeed-sdk : Add systemd.hostname as kernel
 parameter

Platform based Hostname is created at the u-boot and passed as bootarg,
systemd picks it up and sets the proper hostname at the very
early boot process to avoid the Network confusion.

Signed-off-by: Vinu Vaghasia <vinu.vaghasia@amd.com>
---
 common/main.c | 96 ++++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 95 insertions(+), 1 deletion(-)

diff --git a/common/main.c b/common/main.c
index b5e119b203..78dd743096 100644
--- a/common/main.c
+++ b/common/main.c
@@ -13,6 +13,7 @@
 #include <version.h>
 #include <i2c.h>
 #include <linux/ctype.h>
+#include <string.h>

 #define EEPROM_I2C_BUS			7
 #define EEPROM_DEV_ADDR		0X50
@@ -21,11 +22,104 @@
 #define BOARD_ID_LEN			1
 #define BOARD_ID_BUFF_LEN		2

+#define ONYX_SLT        61  //0x3D
+#define ONYX_1          64  //0x40
+#define ONYX_2          65  //0x41
+#define ONYX_3          66  //0x42
+#define ONYX_FR4        82  //0x52
+#define QUARTZ_DAP      62  //0x3E
+#define QUARTZ_1        67  //0x43
+#define QUARTZ_2        68  //0x44
+#define QUARTZ_3        69  //0x45
+#define QUARTZ_FR4      81  //0x51
+#define RUBY_1          70  //0x46
+#define RUBY_2          71  //0x47
+#define RUBY_3          72  //0x48
+#define TITANITE_1      73  //0x49
+#define TITANITE_2      74  //0x4A
+#define TITANITE_3      75  //0x4B
+#define TITANITE_4      76  //0x4C
+#define TITANITE_5      77  //0x4D
+#define TITANITE_6      78  //0x4E
+
+#define HOSTNAME_BUFF_LEN       32
+#define CMD_BOOT_ARGS_BUFF_LEN  512
+#define HEX_BASE                16
+#define OCTATE_5_OFFSET         12
+#define OCTATE_6_OFFSET         15
+#define MAC_ADDR_ARRAY_SIZE     17
+
 /*
  * Board-specific Platform code can reimplement show_boot_progress () if needed
  */
 __weak void show_boot_progress(int val) {}

+/*
+ * Read board_id env, Read ethaddr env
+ * set the hostname as Kernel cmd args
+ */
+void set_hostname()
+{
+    char *board_id;
+    char *ethaddr;
+    char hostname[HOSTNAME_BUFF_LEN];
+    char *cur_bootargs;
+    char new_bootargs[CMD_BOOT_ARGS_BUFF_LEN];
+    int octate_5, octate_6;
+
+    /*  Read Kernel cmd line args */
+    cur_bootargs =  env_get("bootargs");
+
+    /*  Check whether 'bootargs' alrady contains hostname parameter */
+    if  ( !(strstr(cur_bootargs, "systemd.hostname")) )
+    {
+        /* Read the board_id from   env */
+        board_id = env_get("board_id");
+        /* Read the eth0 MAC address from env */
+        ethaddr = env_get("ethaddr");
+
+        if ( board_id != NULL && ethaddr != NULL && strlen(ethaddr) >= MAC_ADDR_ARRAY_SIZE)
+        {
+            octate_5 = (int)simple_strtol(&ethaddr[OCTATE_5_OFFSET], NULL, HEX_BASE);
+            octate_6 = (int)simple_strtol(&ethaddr[OCTATE_6_OFFSET], NULL, HEX_BASE);
+
+            switch ((int)simple_strtol(board_id, NULL, HEX_BASE))
+            {
+                case ONYX_SLT:
+                case ONYX_FR4:
+                case ONYX_1 ... ONYX_3:
+                    sprintf(hostname, "onyx-%02x%02x", octate_5, octate_6);
+                    break;
+                case QUARTZ_DAP:
+                case QUARTZ_FR4:
+                case QUARTZ_1 ... QUARTZ_3:
+                    sprintf(hostname, "quartz-%02x%02x", octate_5, octate_6);
+                    break;
+                case RUBY_1 ... RUBY_3:
+                    sprintf(hostname, "ruby-%02x%02x", octate_5, octate_6);
+                    break;
+                case TITANITE_1 ... TITANITE_6:
+                    sprintf(hostname, "titanite-%02x%02x", octate_5, octate_6);
+                    break;
+                default:
+                    sprintf(hostname, "sp5-%02x%02x", octate_5, octate_6);
+                    break;
+            }
+            /* Append bootargs with hostname */
+            sprintf(new_bootargs, "%s systemd.hostname=%s", cur_bootargs, hostname);
+            env_set("bootargs", new_bootargs);
+            env_save();
+            printf("Hostname set to '%s'\n", hostname);
+        }
+        else
+            printf("Failed setting hostname, board_id or ethaddr is not set\n");
+    }
+    else
+        printf("Hostname is already set with bootargs\n");
+
+    return;
+}
+
 /*
  * Read board ID from mother board eeprom and set as an env
  */
@@ -102,7 +196,7 @@ void main_loop(void)

 	/* Read board id from eerpom and set env */
 	set_board_id();
-
+	set_hostname();
 	s = bootdelay_process();
 	if (cli_process_fdt(&s))
 		cli_secure_boot_cmd(s);
--
2.17.1

